File: fx/src/main/java/gargoyle/fx/FXRegistry.java
package gargoyle.fx;

import java.io.Closeable;
import java.util.Collection;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArraySet;

@SuppressWarnings({"rawtypes", "unchecked"})
final class FXRegistry implements Closeable {

    private final Collection<FXBeanDef> beanDefs = new CopyOnWriteArraySet<>();

    public FXRegistry() {
    }

    @Override
    public void close() {
        beanDefs.forEach(FXBeanDef::close);
        beanDefs.clear();
    }

    public <T> T getBean(final String name, final Class<T> type) {
        return beanDefs.stream()
                .filter(beanDef -> beanDef.isDeclared(type))
                .filter(beanDef -> beanDef.isNamed(name))
                .map(beanDef -> ((FXBeanDef<T, T>) beanDef).get())
                .collect(FXUtil.toSingleton(() -> new FXException(FXUtil.format("no bean {name}",
                        Map.of("name", name)))));
    }

    public <T> T getBean(final Class<T> type) {
        return beanDefs.stream()
                .filter(beanDef -> beanDef.isDeclared(type))
                .map(beanDef -> ((FXBeanDef<T, T>) beanDef).get())
                .collect(FXUtil.toSingleton(() -> new FXException("no bean " + type.getName())));
    }

    public <D, I extends D> void register(final String name,
                                          final Class<D> declarationClass, final Class<I> implementationClass) {
        final boolean added = beanDefs.add(new FXBeanDef<>(name, declarationClass, implementationClass, null));
        if (!added) throw new FXException(FXUtil.format("{name} already registered", Map.of("name", name)));
    }

    public <D, I extends D> void register(final String name,
                                          final Class<D> declarationClass, final Class<? extends I> implementationClass,
                                          final I existingBean) {
        final boolean added = beanDefs.add(new FXBeanDef<>(name, declarationClass, implementationClass, existingBean));
        if (!added) throw new FXException(FXUtil.format("{name} already registered", Map.of("name", name)));
    }

    @Override
    public String toString() {
        return FXUtil.format("FXRegistry{defs={array}}", Map.of("array", beanDefs));
    }
}

File: fx/src/main/kotlin/gargoyle/fx/FxRegistry.kt
package gargoyle.fx

import gargoyle.fx.log.FxLog
import java.io.Closeable
import java.io.IOException
import java.util.concurrent.CopyOnWriteArraySet
import kotlin.reflect.KClass
import kotlin.reflect.full.isSubclassOf

class FxRegistry : Closeable {
    private val beans: MutableSet<FxBeanInfo<*, *>> = CopyOnWriteArraySet()

    override fun close(): Unit =
        with(beans) {
            forEach { it.close() }
            clear()
        }

    fun <T : Any> getBean(name: String, type: KClass<T>): T =
        findBeanInfo<T, T> { it.isDeclared(type) && it.isNamed(name) }?.let { it() } ?: error("no bean $name")

    fun <T : Any> getBean(type: KClass<T>): T =
        findBeanInfo<T, T> { it.isDeclared(type) }?.let { it() } ?: error("no bean ${type.qualifiedName}")

    fun <D : Any, I : D> register(name: String, declaration: KClass<D>, implementation: KClass<I>): Unit =
        FxBeanInfo(declaration, implementation, name)
            .let { if (it !in beans) beans += it else error("$name already registered") }

    fun <D : Any, I : D> register(
        name: String,
        declaration: KClass<D>,
        implementation: KClass<I>,
        existingBean: I
    ): Unit =
        FxBeanInfo(declaration, implementation, name, lazy { existingBean })
            .let { if (it !in beans) beans += it else error("$name already registered") }

    override fun toString(): String =
        "FXRegistry{beans=$beans}"

    @Suppress("UNCHECKED_CAST")
    private fun <D : Any, I : D> findBeanInfo(predicate: (FxBeanInfo<*, *>) -> Boolean): FxBeanInfo<D, I>? =
        beans.filter(predicate).map { (it as FxBeanInfo<D, I>) }.firstOrNull()

    internal class FxBeanInfo<D : Any, I : D>(
        private val declaration: KClass<D>,
        private val implementation: KClass<I>,
        private val name: String = declaration.qualifiedName!!,
        private var instance: Lazy<I> = lazy { FxIntUtil.newInstance(implementation) }
    ) : Comparable<FxBeanInfo<D, I>>, () -> I, Closeable {
        private val lock = Any()
        private val instantiated
            get() = instance.isInitialized()

        override fun close() {
            synchronized(lock) {
                if (instantiated && instance.value is Closeable)
                    try {
                        (instance.value as Closeable).close()
                    } catch (e: IOException) {
                        FxLog.error("$this close", e)
                    }
            }
        }

        override fun compareTo(other: FxBeanInfo<D, I>): Int =
            name.compareTo(other.name)

        override fun invoke(): I =
            synchronized(lock) { instance.value }

        override fun hashCode(): Int =
            name.hashCode()

        override fun equals(other: Any?): Boolean =
            this === other || null != other && javaClass == other.javaClass && name == (other as FxBeanInfo<*, *>).name

        override fun toString(): String =
            "FXBeanDef{name=$name, instantiated=$instantiated, declaration=$declaration, implementation=$implementation}"

        fun <T : Any> isDeclared(type: KClass<T>): Boolean =
            type.isSubclassOf(declaration)

        fun isNamed(otherName: String): Boolean =
            otherName == name
    }
}
